docs/book/v2/quick-start.md

```diff
@@ -0,0 +1,149 @@
+
+## Writing Middleware
+
+The next step is writing actual middleware to dispatch. PSR-15 defines two different interfaces:
+
+### `MiddlewareInterface` vs. `RequestHandlerInterface`
+
+_Middleware_ is code sitting between a request and a response; it typically analyzes the request to aggregate incoming data, delegates it to another layer to process, and then creates and returns a response.
+
+A _RequestHandler_ is a class that receives a request and returns a response, without delegating to other layers of the application. This is generally the inner-most layer of your application.
+
+For more in-depth documentation visit the documentation for [Mezzio](https://docs.mezzio.dev/mezzio/v3/getting-started/features/)
+and [Stratigility](https://docs.laminas.dev/laminas-stratigility/v3/intro/) or the [PSR specification](https://www.php-fig.org/psr/psr-15/).
+
+### Request Handlers
+
+```php
+namespace Application\Handler;
+
+use Application\Entity\Album;
+use Laminas\Diactoros\Response\TextResponse;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class AlbumDetailMiddleware implements RequestHandlerInterface
+{
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        /** @var Album $album */
+        $album = $request->getAttribute('album');
+
+        return new TextResponse('The name of the album is: ' . $album->getName());
+    }
+}
+```
+
+RequestHandlers resemble a single MVC Controller action, so mostly a `ReqeustHandler` will be used to dispatch a request.
+
+### Middleware
+
+```php
+namespace Application\Middleware;
+
+use Application\Repository\AlbumRepositoryInterface;
+use Fig\Http\Message\StatusCodeInterface;
+use Laminas\Diactoros\Response\TextResponse;
+use Laminas\Router\RouteMatch;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class AlbumFromRouteMiddleware implements MiddlewareInterface
+{
+    /** @var AlbumRepositoryInterface */
+    private $albumRepository;
+
+    public function __construct(AlbumRepositoryInterface $albumRepository)
+    {
+        $this->albumRepository = $albumRepository;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        /** @var RouteMatch $routeMatch */
+        $routeMatch = $request->getAttribute(RouteMatch::class);
+        $albumId    = $routeMatch->getParam('album_id');
+        $album      = $this->albumRepository->findById($albumId);
+
+        // if no album was found, we short-circuit the pipe and return a 404 error:
+        if ($album === null) {
+            return new TextResponse(
+                sprintf('Album with ID %s not found!', $albumId),
+                StatusCodeInterface::STATUS_NOT_FOUND
+            );
```

Comment:
Please don't use a specific response implementation! (also,  you've not imported this class anyways!).

Instead, compose a PSR-17 response factory via the constructor, and use the repsonse factory to create the response. This helps decouple from a specific PSR-7 implementation.

Reply:
I _did_ import it. But you're right, I've rewritten it to use ResponseFactory and StreamFactory, instead.